### ISSUE TITLE
Explore possibilities for modular implicits

### ISSUE THREAD

**Issue by bjz** [Marwes/gluon#47]
At the moment module polymorphism is kind of clunky. Whilst our priority is more focused on polishing what we already have, it might be good to note down our thoughts here.

**Comment by bjz** [comment]
Here is an idea that uses extensible records to implement module inheritance:

```ocaml
type (f: Functor) =
    { t : Type -> Type -> Type
    , map : [a, b] -> (a -> b) -> f.t a -> f.t b
    }

type (f: Applicative) =
    { Functor
    | apply : [a, b] -> f.t (a -> b) -> f.t a -> f.t b
    , const : [a, b] -> f.t a -> f.t b -> f.t b
    }

type (m : Monad) =
    { Applicative
    | flat_map : [a, b] -> (a -> m.t b) -> m.t a -> m.t b
    }

let (>>=) : [m : Monad] -> m.t a -> (a -> m.t b) -> m.t b
    (>>=) [m] = flip m.flat_map
```

**Comment by bjz** [comment]
One issue with the above is how to handle multi-parameter inheritance.

Maybe:

```ocaml
type (m : Monad) =
    { foldable : Foldable
    , applicative : Applicative
    , t : foldable.t & applicative.t
    , flat_map : [a, b] -> (a -> m.t b) -> m.t a -> m.t b
    }
```

**Comment by bjz** [comment]
One issue with that last approach is traversing up the hierarchy to get to the function you want:

```haskell
foo.monad.applicative.functor.map f xs
```

blegh

**Comment by Marwes** [comment]
```f#
m.applicative.functor.map f xs
```
Yeah that does not look fun. Row polymorphism might be useful here which purescript uses.

```f#
type Functor f = {
    map : (a -> b) -> f a -> f b,
    | r
}

type Applicative f = {
    map : (a -> b) -> f a -> f b,
    (<*>) : f (a -> b) -> f a -> f b,
    pure : a -> f a,
    | r
}
```

This would mean any `Àpplicative` is also a `Functor`.

With the ability to 'splat' records it might be pretty easy to write.

```f#
let applicative_list = {
    (<*>) = ..,
    pure = ...,
    .. functor_list // splat the functor instance into this record
}
```

https://leanpub.com/purescript/read#leanpub-auto-record-patterns-and-row-polymorphism

**Comment by bjz** [comment]
Your need another type parameter in your examples. ie:

```haskell
type Functor (f : Type -> Type) (r : Row) = {
    map : (a -> b) -> f a -> f b,
    | r
}
```

**Comment by bjz** [comment]
Or did you mean:

```haskell
type Functor (f : Type -> Type) = forall (r : Row) . {
    map : (a -> b) -> f a -> f b,
    | r
}
```

**Comment by Marwes** [comment]
Actually I believe it should be.

```
type Functor f = forall r. {
    map : (a -> b) -> f a -> f b,
    | r
}
```

**Comment by bjz** [comment]
Interesting stuff in the [1ML paper](https://www.mpi-sws.org/~rossberg/1ml/1ml-extended.pdf) - in some respects it looks strikingly similar to what you have arrived at in Gluon:

```sml
type MONAD (m : type => type) = {
    return a : a -> m a;
    bind a b : m a -> (a - > m b) -> m b;
};

map a b (m : type => type) (M : MONAD m) (f : a -> b) (mx : m a) =
    M.bind a b mx (fun (x : a) => M.return b (f x))
```

**Comment by bjz** [comment]
One interesting syntactic thing is that:

```sml
type FOO (a : type) = { ... };
```

Is sugar for:

```sml
FOO = (a : type) -> type { ... };
```

So:

```sml
type EQ = { type t; eq : t -> t };
```

is the same as:

```sml
type EQ = { t : type; eq : t -> t };
```

Giving the following insight:

> The record type `EQ` amounts to a module signature, since it contains an abstract type component `t`. It is referred to in the type of `eq`, which shows that record types are seemingly “dependent”: like for terms, earlier components are in scope for later components – the key insight of the F-ing approach is that this dependency is benign, however, and can be translated away, as we will see in Section 3.

**Comment by Marwes** [comment]
Definitely need to re-read the 1ML paper. While it does have some similarities gluon is much simpler currently as types in records can only be a single type.

```f#
type Test a = {
    F: Int -> Int // `F` is a unqiue type whereas in 1ML or Rust it may depend on `a`
}
```

I never quite got why 1ML allowed types to be used as values (or I did get it and forgot). Types shouldn't appear in the actual code that is executed (so they aren't values during actual execution) and you don't actually want to be able to pass around and modify types arbitrarily in a Turing complete language anyway as that makes inference impossible

**Comment by bjz** [comment]
I don't really see that as a problem tbh - in fact it makes the language simpler by removing the distinction between types and values and seems to get rid of a separate kind language. 1ML restricts this somewhat in order to get good type inference, as described in Section 2, under 'Predicativity'.

**Comment by Marwes** [comment]
There are probably to many useful constructs that are impossible if we take a hard stance on preserving decidability so I am not to attached to having a decidable type system. As long as inference is predictable and is able to give good errors it is fine to require some type signatures to help out. I'd even go a bit further to say that all exported bindings should have a type signature (which could be checked with a lint defaulting to error when a signature is missing).

**Comment by bjz** [comment]
Yeah - I think having good errors is critical. That is one of my concerns with too much language simplification and leaning on inference - the increased expressiveness may make it harder to provide useful feedback to the user.

**Comment by bjz** [comment]
So it seems that the language already has the ability to express interface inheritance! 😲 

```diff
 type Applicative f = {
-    (<*>) : f (a -> b) -> f a -> f b,
+    functor : Functor f,
+    apply : f (a -> b) -> f a -> f b,
     pure : a -> f a
 }
```

(see  #56)

**Comment by brendanzab** [comment]
Discussion of modular typeclasses for Purescript: purescript/purescript#1886

**Comment by Marwes** [comment]
Implicits in Agda https://github.com/bitonic/tog/wiki/Implicit-Arguments

**Comment by brendanzab** [comment]
Note that Agda calls their implicits 'instance arguments' (as opposed to 'implicit arguments'). But the stuff about unifying named arguments could still be relevant.
